> * 실습 데이터:  2021년 2월‘서울시 지하철호선별 역별 승하차 인원 정보’(CARD_SUBWAY_MONTH_202102.csv)

In [7]:
df = pd.read_csv('CARD_SUBWAY_MONTH_202102.csv', index_col= False)
df

사용일자  노선명   역명  승차총승객수  하차총승객수      등록일자
0      20210201  중앙선   용문    1282    1259  20210204
1      20210201  중앙선   지평      48      43  20210204
2      20210201  중앙선   아신     415     413  20210204
3      20210201  중앙선   양수    1463    1466  20210204
4      20210201  중앙선  운길산     495     463  20210204
...         ...  ...  ...     ...     ...       ...
16746  20210228  중앙선   오빈     216     236  20210303
16747  20210228  중앙선   양평    3299    3187  20210303
16748  20210228  중앙선   원덕     322     304  20210303
16749  20210228  중앙선   용문    1899    1716  20210303
16750  20210228  중앙선   지평      30      31  20210303

[16751 rows x 6 columns]

# 3.2.6 삭제

<div class="alert alert-success">
    [문제]  
    
    
데이터프레임에서 0번 로우를 삭제하세요.
</div>

In [31]:
df.drop(0)

사용일자  노선명   역명  승차총승객수  하차총승객수      등록일자
1      20210201  중앙선   지평      48      43  20210204
2      20210201  중앙선   아신     415     413  20210204
3      20210201  중앙선   양수    1463    1466  20210204
4      20210201  중앙선  운길산     495     463  20210204
5      20210201  중앙선   팔당     687     662  20210204
...         ...  ...  ...     ...     ...       ...
16746  20210228  중앙선   오빈     216     236  20210303
16747  20210228  중앙선   양평    3299    3187  20210303
16748  20210228  중앙선   원덕     322     304  20210303
16749  20210228  중앙선   용문    1899    1716  20210303
16750  20210228  중앙선   지평      30      31  20210303

[16750 rows x 6 columns]

<div class="alert alert-success">
    [문제]  
    
    
데이터프레임에서 ‘등록일자’ 컬럼을 삭제하세요
</div>

사용일자  노선명   역명  승차총승객수  하차총승객수
0      20210201  중앙선   용문    1282    1259
1      20210201  중앙선   지평      48      43
2      20210201  중앙선   아신     415     413
3      20210201  중앙선   양수    1463    1466
4      20210201  중앙선  운길산     495     463
...         ...  ...  ...     ...     ...
16746  20210228  중앙선   오빈     216     236
16747  20210228  중앙선   양평    3299    3187
16748  20210228  중앙선   원덕     322     304
16749  20210228  중앙선   용문    1899    1716
16750  20210228  중앙선   지평      30      31

[16751 rows x 5 columns]

# 3.2.7 데이터프레임 연산

## ① 산술연산

<div class="alert alert-success">
    [문제]  
    
‘승차총승객수’ 컬럼의 모든 값을 2씩 증가시키세요.
</div>

0        1284
1          50
2         417
3        1465
4         497
         ... 
16746     218
16747    3301
16748     324
16749    1901
16750      32
Name: 승차총승객수, Length: 16751, dtype: int64

<div class="alert alert-success">
    [문제]  
    
'승차총승객수’와 ‘하차총승객수’ 컬럼의 차이를 계산한 ‘승하차총승객수차이’ 컬럼을 생성하세요.
</div>

사용일자  노선명   역명  승차총승객수  하차총승객수      등록일자  승하차총승객수차이
0  20210201  중앙선   용문    1282    1259  20210204         23
1  20210201  중앙선   지평      48      43  20210204          5
2  20210201  중앙선   아신     415     413  20210204          2
3  20210201  중앙선   양수    1463    1466  20210204         -3
4  20210201  중앙선  운길산     495     463  20210204         32

<div class="alert alert-success">
    [문제]  
    
‘승하차총승객수차이’의 평균값을 계산하세요
</div>

30.870276401408873

## ② 샘플링

<div class="alert alert-success">
    [문제]  
    
 데이터프레임에서 10개의 아이템을 샘플링하세요.
</div>

In [36]:
#비복원추출


사용일자  노선명              역명  승차총승객수  하차총승객수      등록일자  승하차총승객수차이
5010   20210209  경부선             영등포   33740   35066  20210212      -1326
8240   20210214  3호선           고속터미널   46854   27780  20210217      19074
15568  20210227  2호선  동대문역사문화공원(DDP)    9395    9940  20210302       -545
8030   20210214  9호선             여의도    3982    4045  20210217        -63
10479  20210218  경부선              금정   23957   24405  20210221       -448
14593  20210225  7호선      숭실대입구(살피재)   11126   11339  20210228       -213
8192   20210214  5호선             까치산   11835   12333  20210217       -498
16638  20210228  분당선              도곡    2221    2114  20210303        107
2669   20210205  경부선             오산대    2802    2527  20210208        275
8065   20210214  7호선              상동    4899    4729  20210217        170

## ③ 값 치환

<div class="alert alert-success">
    [문제]  
    
샘플링된 데이터프레임(sample_df)의 값에서 ‘노선명’ 컬럼의 값이 1호선, 2호선, 3호선 4호선인 
값을 영어(line1, line2, line3, line4)로 변경하세요
</div>

5010       경부선
8240     line3
15568    line2
8030       9호선
10479      경부선
14593      7호선
8192       5호선
16638      분당선
2669       경부선
8065       7호선
Name: 노선명, dtype: object

## ④ 함수 적용

<div class="alert alert-success">
    [문제]  
데이터프레임(df)의 ‘사용일자’ 컬럼을 연도-월-일(예:2021-02-01)의 형태로 표시되도록 수정하
세요.
</div>

'2021-02-01'

0        2021-02-01
1        2021-02-01
2        2021-02-01
3        2021-02-01
4        2021-02-01
            ...    
16746    2021-02-28
16747    2021-02-28
16748    2021-02-28
16749    2021-02-28
16750    2021-02-28
Name: 사용일자, Length: 16751, dtype: object

<div class="alert alert-success">
    [문제]  
데이터프레임(df)의 ‘사용일자1’ 컬럼을 새로 생성하고, ‘사용일자’ 컬럼에 있는 값을 datetime64의 
타입으로 변환하여 넣으세요.
</div>

사용일자  노선명   역명  승차총승객수  하차총승객수      등록일자  승하차총승객수차이       사용일자1
0  20210201  중앙선   용문    1282    1259  20210204         23  2021-02-01
1  20210201  중앙선   지평      48      43  20210204          5  2021-02-01
2  20210201  중앙선   아신     415     413  20210204          2  2021-02-01
3  20210201  중앙선   양수    1463    1466  20210204         -3  2021-02-01
4  20210201  중앙선  운길산     495     463  20210204         32  2021-02-01

0       2021-02-01
1       2021-02-01
2       2021-02-01
3       2021-02-01
4       2021-02-01
           ...    
16746   2021-02-28
16747   2021-02-28
16748   2021-02-28
16749   2021-02-28
16750   2021-02-28
Name: 사용일자1, Length: 16751, dtype: datetime64[ns]

------

## ⑤ 원핫인코딩

노선_2호선  노선_3호선  노선_5호선  노선_7호선  노선_9호선  노선_경부선  노선_분당선
5010        0       0       0       0       0       1       0
8240        0       1       0       0       0       0       0
15568       1       0       0       0       0       0       0
8030        0       0       0       0       1       0       0
10479       0       0       0       0       0       1       0
14593       0       0       0       1       0       0       0
8192        0       0       1       0       0       0       0
16638       0       0       0       0       0       0       1
2669        0       0       0       0       0       1       0
8065        0       0       0       1       0       0       0

## ⑥ 결측치 처리

<div class="alert alert-success">
    [문제]  
결측치가 포함된 데이터프레임을 생성하세요.
</div>

In [47]:
import numpy as np
data = {'학번':range(2000,2010),
    '국어성적': [85, 95, 75, 70, 100, np.nan, 95, 85, 80, 85],
    '영어성적': [95, 70, 100, 85, 80, 85, 95, 95, np.nan, 70],
    '수학성적': [np.nan, np.nan, np.nan, np.nan, np.nan, np.nan, np.nan, 95, np.nan, 70],   }
df = pd.DataFrame(data)
df

학번   국어성적   영어성적  수학성적
0  2000   85.0   95.0   NaN
1  2001   95.0   70.0   NaN
2  2002   75.0  100.0   NaN
3  2003   70.0   85.0   NaN
4  2004  100.0   80.0   NaN
5  2005    NaN   85.0   NaN
6  2006   95.0   95.0   NaN
7  2007   85.0   95.0  95.0
8  2008   80.0    NaN   NaN
9  2009   85.0   70.0  70.0

<div class="alert alert-success">
    [문제]  
위에서 생성한 데이터프레임에 결측치가 포함되어 있는지 확인하는 코드를 작성하세요.
</div>

학번   국어성적   영어성적   수학성적
0  False  False  False   True
1  False  False  False   True
2  False  False  False   True
3  False  False  False   True
4  False  False  False   True
5  False   True  False   True
6  False  False  False   True
7  False  False  False  False
8  False  False   True   True
9  False  False  False  False

In [49]:
df.isna().sum()

학번      0
국어성적    1
영어성적    1
수학성적    8
dtype: int64

<div class="alert alert-success">
    [문제]  
데이터프레임에 포함된 모든 결측치를 0으로 치환하세요.
</div>

학번   국어성적   영어성적  수학성적
0  2000   85.0   95.0   0.0
1  2001   95.0   70.0   0.0
2  2002   75.0  100.0   0.0
3  2003   70.0   85.0   0.0
4  2004  100.0   80.0   0.0
5  2005    0.0   85.0   0.0
6  2006   95.0   95.0   0.0
7  2007   85.0   95.0  95.0
8  2008   80.0    0.0   0.0
9  2009   85.0   70.0  70.0

In [51]:
df['수학성적'].fillna(50, inplace=True)
df

학번   국어성적   영어성적  수학성적
0  2000   85.0   95.0  50.0
1  2001   95.0   70.0  50.0
2  2002   75.0  100.0  50.0
3  2003   70.0   85.0  50.0
4  2004  100.0   80.0  50.0
5  2005    NaN   85.0  50.0
6  2006   95.0   95.0  50.0
7  2007   85.0   95.0  95.0
8  2008   80.0    NaN  50.0
9  2009   85.0   70.0  70.0

<div class="alert alert-success">
    [문제]  
데이터프레임에 포함된 모든 결측치를 삭제하세요.
</div>

학번   국어성적   영어성적  수학성적
0  2000   85.0   95.0  50.0
1  2001   95.0   70.0  50.0
2  2002   75.0  100.0  50.0
3  2003   70.0   85.0  50.0
4  2004  100.0   80.0  50.0
6  2006   95.0   95.0  50.0
7  2007   85.0   95.0  95.0
9  2009   85.0   70.0  70.0

# 3.2.8 변형

<div class="alert alert-success">
    [문제]  
 ‘서울시 지하철호선별 역별 승하차 인원 정보’의 데이터프레임을 ‘노선명’으로 그룹핑하고, 각 필드
의 데이터의 평균값을 표시하도록 코드를 작성하세요.
</div>

사용일자        승차총승객수        하차총승객수          등록일자
노선명                                                             
1호선       2.021021e+07  16711.396429  16224.032143  2.021023e+07
2호선       2.021021e+07  20258.510000  20406.605714  2.021023e+07
3호선       2.021021e+07  11563.884739  11542.535752  2.021023e+07
4호선       2.021021e+07  13839.026099  13953.564560  2.021023e+07
5호선       2.021021e+07   8774.492588   8727.040431  2.021023e+07
6호선       2.021021e+07   6499.918173   6405.854424  2.021023e+07
7호선       2.021021e+07  10210.968487  10060.233193  2.021023e+07
8호선       2.021021e+07   8356.453782   8425.422269  2.021023e+07
9호선       2.021021e+07   8110.387143   8246.852857  2.021023e+07
9호선2~3단계  2.021021e+07   5238.137363   5156.379121  2.021023e+07
경강선       2.021021e+07   1970.116883   1908.357143  2.021023e+07
경부선       2.021021e+07   8366.500000   8366.444139  2.021023e+07
경원선       2.021021e+07   5051.796590   4940.018270  2.021023e+07
경의선       2.021021e+07   2962.963365   2910.443691  2.021023e+07
경인선       2.021021e+07  10789.323214  10672.037500  2.021023e+07
경춘선       2.021021e+07   1462.635338   1409.640977  2.021023e+07
공항철도 1호선  2.021021e+07   5076.122449   4643.492347  2.021023e+07
과천선       2.021021e+07   8984.544643   8827.571429  2.021023e+07
분당선       2.021021e+07   7849.137539   8079.867632  2.021023e+07
수인선       2.021021e+07   2341.289683   2335.438492  2.021023e+07
안산선       2.021021e+07   6966.684066   6926.851648  2.021023e+07
우이신설선     2.021021e+07   2559.425824   2508.461538  2.021023e+07
일산선       2.021021e+07   8072.929078   7861.592199  2.021023e+07
장항선       2.021021e+07   1495.559524   1447.095238  2.021023e+07
중앙선       2.021021e+07   3235.462585   3173.188776  2.021023e+07

<div class="alert alert-success">
    [문제]  
다음 두 개의 데이터프레임을 하나의 데이터프레임으로 합쳐보세요.
    
| 이름 | 성별 |
| ---- | ---- |
| A    | 남자 |
| B    | 여자 |
| D    | 남자 |



| 이름 | 키   |
| ---- | ---- |
| A    | 180  |
| B    | 160  |
| C    | 150  |


</div>

이름  성별
0  A  남자
1  B  여자
2  D  남자

이름    키
0  A  180
1  B  160
2  C  150

이름  성별    키
0  A  남자  180
1  B  여자  160

이름  성별      키
0  A  남자  180.0
1  B  여자  160.0
2  D  남자    NaN

이름   성별    키
0  A   남자  180
1  B   여자  160
2  C  NaN  150

이름   성별      키
0  A   남자  180.0
1  B   여자  160.0
2  D   남자    NaN
3  C  NaN  150.0

<div class="alert alert-success">
    [문제]  
다음 두 개의 데이터프레임을 하나의 데이터프레임으로 합쳐보세요.
    
|   이름 |   키 | 나이 | 주소 |
| ---: | ---: | ---: | ---- |
|    A |  180 |   23 | 서울 |
|    B |  160 |   46 | 부산 |
|    C |  150 |   33 | 대전 |
|    D |  170 |   65 | 원주 |
|    E |  170 |   28 | 제주 |

| 이름 |   키 | 나이 | 주소 |
| ---: | ---: | ---: | ---: |
|    F |  190 |   22 | 전주 |
|    G |  180 |   18 | 서울 |
|    H |  150 |   87 | 수원 |
|    I |  160 |   45 | 부산 |
|    J |  170 |   74 | 인천 |
</div>

이름    키  나이  주소
0  A  180  23  서울
1  B  160  46  부산
2  C  150  33  대전
3  D  170  65  원주
4  E  170  28  제주

이름    키  나이  주소
0  F  190  22  전주
1  G  180  18  서울
2  H  150  87  수원
3  I  160  45  부산
4  J  170  74  인천

이름    키  나이  주소
0  A  180  23  서울
1  B  160  46  부산
2  C  150  33  대전
3  D  170  65  원주
4  E  170  28  제주
0  F  190  22  전주
1  G  180  18  서울
2  H  150  87  수원
3  I  160  45  부산
4  J  170  74  인천

이름    키  나이  주소 이름    키  나이  주소
0  A  180  23  서울  F  190  22  전주
1  B  160  46  부산  G  180  18  서울
2  C  150  33  대전  H  150  87  수원
3  D  170  65  원주  I  160  45  부산
4  E  170  28  제주  J  170  74  인천

# 3.2.9 저장

이름  성별
0  A  남자
1  B  여자
2  D  남자

0  A  남자
0  1  B  여자
1  2  D  남자

A  남자
0       
1  B  여자
2  D  남자